src/CoreWCF.Primitives/src/CoreWCF/Channels/ServiceChannel.cs

```diff
@@ -709,14 +709,20 @@ internal void DecrementActivity()
             }
 
             if (updatedActivityCount == 0 && _autoClose)
+            {
+                return AutoCloseAsync();
+            }
```

Comment:
Can you lift the `if (State == CommunicationState.Opened)` check to be around the call to AutoCloseAsync instead? I understand that you made a private inline method to avoid the async state machine and Task allocation when closing if it's not needed. Moving the if block to here will help with that more.